src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -45,5 +32,212 @@ private static IEnumerable<TSource> PrependIterator<TSource>(IEnumerable<TSource
+            private readonly IEnumerable<TSource> _third;
+
+            internal Concat3Iterator(IEnumerable<TSource> first, IEnumerable<TSource> second, IEnumerable<TSource> third)
+            {
+                Debug.Assert(first != null && second != null && third != null);
+                _first = first;
+                _second = second;
+                _third = third;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new Concat3Iterator<TSource>(_first, _second, _third);
+            }
+
+            internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
+            {
+                return new ConcatNIterator<TSource>(this, next, 3);
+            }
+
+            internal override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                switch (index)
+                {
+                    case 0: return _first;
+                    case 1: return _second;
+                    case 2: return _third;
+                    default: return null;
+                }
+            }
+        }
+
+        private sealed class ConcatNIterator<TSource> : ConcatIterator<TSource>
+        {
+            // To handle chains of >= 4 sources, we chain the concat iterators together and allow
+            // GetEnumerable to fetch enumerables from the previous sources.  This means that rather
+            // than each MoveNext/Current calls having to traverse all of the previous sources, we
+            // only have to traverse all of the previous sources once per chained enumerable.  An
+            // alternative would be to use an array to store all of the enumerables, but this has
+            // a much better memory profile and without much additional run-time cost.
+
+            private readonly ConcatIterator<TSource> _previousConcat;
+            private readonly IEnumerable<TSource> _next;
+            private readonly int _nextIndex;
+
+            internal ConcatNIterator(ConcatIterator<TSource> previousConcat, IEnumerable<TSource> next, int nextIndex)
+            {
+                Debug.Assert(previousConcat != null);
+                Debug.Assert(next != null);
+                Debug.Assert(nextIndex > 0);
+                _previousConcat = previousConcat;
+                _next = next;
+                _nextIndex = nextIndex;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new ConcatNIterator<TSource>(_previousConcat, _next, _nextIndex);
+            }
+
+            internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
+            {
+                return new ConcatNIterator<TSource>(this, next, checked(_nextIndex + 1));
+            }
+
+            internal override IEnumerable<TSource> GetEnumerable(int index)
+            {
+                return
+                    index < _nextIndex ? _previousConcat.GetEnumerable(index) :
```

Comment:
> This seems to be O(n^2) on the number of concatenated fragments

My point is, today it's O(n^2) on the number of items in the enumerables (a MoveNext and a Current call per item).  This change makes it O(n^2) on the number of enumerables (a GetEnumerable call per enumerable).  So, yes, for long chains of very short fragments, it could approach with a constant multiple of what it is today, though still much less.

> Would it make sense to memoize the chain into a List if we know the chain is sufficiently long?

That's what I initially had, actually, where ConcatNIterator stored an array of the enumerables rather than a link back to its previous one, but it requires allocating such an array/list, which is why I moved away from it.  Are you asking that I add back such a thing for use with long chains, e.g. use Concat2Iterator for chains of 2, Concat3Iterator for chains of 3, ConcatLinkedIterator for chains of 4-16 (what's currently in this PR called ConcatNIterator), and a new ConcatArrayIterator for chains longer than 16? Or are you suggesting that when enumeration starts, build up an array of the enumerables once and then iterate through that?  

I'm open to doing things like that.  I just want to highlight that what's here in the PR is strictly better than what's currently checked in, at least in this regard.